central/delegatedregistryconfig/service/service.go

```diff
@@ -0,0 +1,227 @@
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	config, err := s.dataStore.GetConfig(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving config")
+	}
+
+	if config == nil {
+		return &storage.DelegatedRegistryConfig{}, nil
+	}
+
+	return config, nil
+}
+
+// GetClusters returns the list of clusters (id + name) that are eligible for delegating scanning
+// requests (ie: only clusters with scanners that understand the delegated registry config)
+func (s *serviceImpl) GetClusters(ctx context.Context, _ *v1.Empty) (*v1.DelegatedRegistryClustersResponse, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	clusters, err := s.getClusters(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving clusters")
+	}
+
+	if len(clusters) == 0 {
+		return nil, status.Error(codes.NotFound, "no valid clusters found")
+	}
+
+	return &v1.DelegatedRegistryClustersResponse{
+		Clusters: clusters,
+	}, nil
+}
+
+// PutConfig updates Central's delegated registry config
+func (s *serviceImpl) PutConfig(ctx context.Context, config *storage.DelegatedRegistryConfig) (*storage.DelegatedRegistryConfig, error) {
+	if s.dataStore == nil {
+		return nil, status.Error(codes.Unimplemented, "datastore not initialized, is postgres enabled?")
+	}
+
+	if err := s.validate(ctx, config); err != nil {
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	if err := s.dataStore.UpsertConfig(ctx, config); err != nil {
+		return nil, errors.Wrap(err, "upserting config")
+	}
+
+	return config, nil
+}
+
+func (s *serviceImpl) validate(ctx context.Context, config *storage.DelegatedRegistryConfig) error {
+	if config == nil {
+		// this block not reachable via GRPC-gateway invocations
+		return errors.New("config missing")
+	}
+
+	if config.EnabledFor == storage.DelegatedRegistryConfig_NONE {
+		// ignore rest of config
+		return nil
+	}
+
+	errorList := errorhelpers.NewErrorList("Validation")
+	// validate the default cluster id
+	if config.DefaultClusterId == "" {
+		errorList.AddStrings("defaultClusterId required if enabledFor is not NONE")
+		return errorList.ToError()
+	}
+
+	// get the valid clusters
+	validClusters, err := s.getValidClusterIds(ctx)
+	if err != nil {
+		return errors.Wrap(err, "unable to validate config")
+	}
+
+	if !validClusters.Contains(config.DefaultClusterId) {
+		errorList.AddStrings(fmt.Sprintf("default cluster %q is not a valid cluster", config.DefaultClusterId))
+	}
+
+	// validate the registries / clusters
+	for i, r := range config.Registries {
+
+		// if a cluster id was provided, check if its valid
+		if r.ClusterId != "" && !validClusters.Contains(r.ClusterId) {
+			errorList.AddStrings(fmt.Sprintf("cluster %q is not valid at index %v", r.ClusterId, i))
```

Comment:
i would omit the at index portion

Reply:
My thought was to make it easier on debugging due to the possibility that multiple 'registries' could be pointing to the same invalid cluster, and the UI would be showing the cluster name not ID. Thinking about it more there are other ways the invalid line items could be identified, so I'll remove the index.

Reply:
Removed in latest commit